### PR TITLE
USD to SDF: Added links

### PR DESCRIPTION
**Pull request by ahcorde**
# 🎉 New feature

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: usd/src/usd_parser/USDLinks.cc [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    _link = new sdf::Link();
+    _link->SetName(_nameLink);
+
+    pxr::UsdPrim tmpPrim = _prim;
+    ignition::math::Pose3d pose;
+    ignition::math::Vector3d scale(1, 1, 1);
+    GetTransform(tmpPrim, _usdData, pose, scale, "");
+    size_t nSlash = std::count(_nameLink.begin(), _nameLink.end(), '/');
+    if (nSlash > 1)
+      _link->SetRawPose(pose);
```
Comment (adlarkin) [inline usd/src/usd_parser/USDLinks.cc]: Why is `nSlash` needed? In the `GetTransform` call, if the prim is the link prim itself, the pose should be `(0, 0, 0, 0, 0, 0)`, right? So, I imagine that we should always be able to set the link's raw pose without any issues.

File: usd/src/usd_parser/USDLinks.cc [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    // _link->SetScale(scale);
```
Comment (adlarkin) [inline usd/src/usd_parser/USDLinks.cc:124]: Can this line be deleted?

File: usd/src/usd_parser/USDLinks.cc [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  // If the schema is a rigid body use this name instead.
+  if (_prim.HasAPI<pxr::UsdPhysicsRigidBodyAPI>())
+  {
+    _link->SetName(primPathStr);
+  }
```
Comment (adlarkin) [inline usd/src/usd_parser/USDLinks.cc]: Why is this needed? Also, should the link's name be set to the basename of the prim path here?

File: usd/src/usd_parser/USDLinks.cc [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    if (nSlash > 1)
+      _link->SetRawPose(pose);
+    // _link->SetScale(scale);
+  }
+
+  // If the schema is a rigid body use this name instead.
+  if (_prim.HasAPI<pxr::UsdPhysicsRigidBodyAPI>())
+  {
+    _link->SetName(pxr::TfStringify(_prim.GetPath()));
+  }
```
Comment (adlarkin) [inline usd/src/usd_parser/USDLinks.cc:131]: Why is this needed? Also, should we be setting the link's name to the basename of the prim path string?

**Review by adlarkin — COMMENTED** [review]
I pushed a few commits with feedback, but I still have a few comments/questions which I left below. There are also a few other things:
1. Would you mind adding tests that show adding links to models works as expected?
2. I see the following warning when I build the code:
```
/sdf_to_usd/src/sdformat/usd/src/usd_parser/USDWorld.cc: In function 'sdf::v12::usd::UsdErrors sdf::v12::usd::parseUSDWorld(const string&, sdf::v12::World&)':
/sdf_to_usd/src/sdformat/usd/src/usd_parser/USDWorld.cc:227:30: warning: 'modelPtr' may be used uninitialized in this function [-Wmaybe-uninitialized]
             modelPtr->AddLink(*link);
```

File: usd/src/usd_parser/USDLinks.cc [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        diagonalInertia = pxr::GfVec3f(0.0001, 0.0001, 0.0001);
+      }
+
+      if (mass < 0.0001)
+      {
+        mass = 0.1;
+      }
+
+      massMatrix.SetMass(mass);
+      massMatrix.SetDiagonalMoments(
```
[the remark this answers is not in the document]
Comment (ahcorde) [reply usd/src/usd_parser/USDLinks.cc:81]: I'm not really sure how to set this values with the current API, any thoughts @adlarkin @scpeters ?

File: usd/src/usd_parser/USDLinks.cc [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  std::string primPathStr = pxr::TfStringify(_prim.GetPath());
+  std::string primType = _prim.GetPrimTypeInfo().GetTypeName().GetText();
+
+  std::pair<std::string, std::shared_ptr<USDStage>> data =
+    _usdData.FindStage(primNameStr);
+
+  // Is this a new link ?
+  if (_link == nullptr)
+  {
+    _link = new sdf::Link();
```
[the remark this answers is not in the document]
Comment (ahcorde) [reply usd/src/usd_parser/USDLinks.cc:115]: we are getting a raw pointer from `sdf::Link` I was trying to keep it simple

File: usd/src/usd_parser/USDLinks.cc (hunk 4 above)
Comment (ahcorde) [reply usd/src/usd_parser/USDLinks.cc:131]: if a path is a rigid body I used this name as the name of the link

**Comment by adlarkin** [comment]
@ahcorde it looks like there are merge conflicts with `sdf12`, would you mind resolving them?

**Comment by ahcorde** [comment]
Done @adlarkin

File: usd/src/usd_parser/USDLinks.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (adlarkin) [reply usd/src/usd_parser/USDLinks.cc:96]: I left a `TODO` note in the code about adding these things in so that we can revisit it later

File: usd/src/usd_parser/USDLinks.cc [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+  const std::string &_nameLink,
+  sdf::Link *_link,
+  USDData &_usdData)
+{
+  std::string primNameStr = _prim.GetPath().GetName();
+  std::string primPathStr = pxr::TfStringify(_prim.GetPath());
+  std::string primType = _prim.GetPrimTypeInfo().GetTypeName().GetText();
+
+  std::pair<std::string, std::shared_ptr<USDStage>> data =
+    _usdData.FindStage(primNameStr);
```
Comment (adlarkin) [inline usd/src/usd_parser/USDLinks.cc:131]: I don't see the `data` variable being used anywhere else in this method. Can this call to `FindStage` be removed? I'm not sure that it's needed.

File: usd/src/usd_parser/USDLinks.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (adlarkin) [reply usd/src/usd_parser/USDLinks.cc]: I changed the API slightly to use `std::optional<sdf::Link>` instead of raw pointers. I think it's a safer approach since it avoids the need for manually allocating/de-allocating memory

File: usd/src/usd_parser/USDLinks.cc [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+  {
+    _link = new sdf::Link();
+    _link->SetName(ignition::common::basename(_nameLink));
+
+    ignition::math::Pose3d pose;
+    ignition::math::Vector3d scale(1, 1, 1);
+    GetTransform(_prim, _usdData, pose, scale, "");
+    size_t nSlash = std::count(_nameLink.begin(), _nameLink.end(), '/');
+    if (nSlash > 1)
+      _link->SetRawPose(pose);
```
Comment (adlarkin) [inline usd/src/usd_parser/USDLinks.cc:144]: @ahcorde I left a [previous comment](https://github.com/ignitionrobotics/sdformat/pull/902#discussion_r840922946) about this, but I still don't understand the purpose of this code. Can you explain why we need to count the slashes and make sure it's `> 1` before setting the pose? I think it would be nice to leave a comment in the code in case this has to be re-visited later.

**Review by adlarkin — COMMENTED** [review]
I pushed a few changes, with the main one being using `std::optional<sdf::Link>` instead of a raw pointer to `sdf::Link`. I have a few final questions below.

File: usd/src/usd_parser/USDLinks.cc (hunk 8 above)
Comment (ahcorde) [reply usd/src/usd_parser/USDLinks.cc:144]: Added a explanation and added TODO

File: usd/src/usd_parser/USDLinks.cc (hunk 7 above)
Comment (ahcorde) [reply usd/src/usd_parser/USDLinks.cc:131]: removed

**Comment by ahcorde** [comment]
@adlarkin ready for a final review

File: usd/src/usd_parser/USDLinks.cc [hunk 9]
```diff
@@ -139,6 +136,10 @@ void ParseUSDLinks(
     ignition::math::Pose3d pose;
     ignition::math::Vector3d scale(1, 1, 1);
     GetTransform(_prim, _usdData, pose, scale, "");
+    // This is a special case when a geometry is defined in the higher level
+    // of the path. we should only set the position is the path at least has
```
Comment (adlarkin) [inline usd/src/usd_parser/USDLinks.cc:140]:
```suggestion
    // of the path. we should only set the position if the path at least has
```

**Review by adlarkin — APPROVED** [review]
